### PR TITLE
Convert GPT-NeoX to HuggingFace

### PR DESCRIPTION
**Pull request by NinedayWang**
This PR includes a script named `convert_neox_pt_to_huggingface_neox.py`, which is used to convert a PolyCoder checkpoint trained by GPT-NeoX into HuggingFace format. A `transformers.GPTNeoXConfig` file that matches the checkpoint is also provided.

I have checked with a 0.4B model, which is trained following the `medium.yml` config. The greedy-decoding outputs by GPT-NeoX's inference script and HF's generate() are identical.

**NOTE:** The HuggingFace model type I used is "gpt_neox", but the model architecture needs to be adjusted. Please refer to [this PR](https://github.com/huggingface/transformers/pull/18695).

### CODE REVIEW

**Comment by VHellendoorn** [comment]
Hi @NinedayWang,

Since it looks like the [HF-side PR](https://github.com/huggingface/transformers/pull/18695) is nearly ready to be merged, I've replicated your PR with the 2.7B checkpoint and am happy to say that things worked pretty much out of the box. 

Concretely, here is what I ran (copied from a larger workflow so there may be a typo or two):

1. To convert the checkpoint: `python3 Convert2HF/convert_neox_pt_to_huggingface_neox.py checkpoints-2-7B/global_step150000 polycoder/2-7B`

    - Note that I modified the `convert` file to take a second command-line argument specifying the output directory *and* to create  any non-existent leading directories in that path. It'd be great if you could make the same changes, and perhaps include a short README explaining the usage of the script. Using `argparse` would be even cleaner since then we get a `help` option for free, but I'll leave that up to you.

2. To load in the model:

```python
from transformers import GPTNeoXForCausalLM, GPTNeoXConfig

converted_dir = './polycoder'
model_name = '2-7B'

ckpt = f'./{converted_dir}/{model_name}'
with open(f'./{converted_dir}/{model_name}-config.json') as f:
    self.config = GPTNeoXConfig(**json.load(f))

model = GPTNeoXForCausalLM.from_pretrained(self.ckpt)
tokenizer = GPT2Tokenizer(f'./{converted_dir}/code-vocab.json', f'./{converted_dir}/code-merges.txt')
```

Here, I am assuming the existence of a `2-7B-config.json` file. It'd be great if you could add that, and the `160M` one, to the PR. They're identical to the 0.4B one except for the number of layers and various dimensions.

3. Query the model:

```python
encoded_input = tokenizer(prompt, return_tensors="pt")
input_ids, attention_mask = encoded_input['input_ids'], encoded_input['attention_mask']
if CUDA_AVAILABLE:
  input_ids = input_ids.cuda()
  attention_mask = attention_mask.cuda()

predictions = model.generate(input_ids=input_ids, attention_mask=attention_mask, do_sample=True, max_length=128, top_p=0.95, temperature=0.2)
print(tokenizer.decode(ids, skip_special_tokens=True, clean_up_tokenization_spaces=False))
```

Naturally, the above has some hyper-parameters that should be changed based on user input. If you want to add instructions such as these to a README with this PR, that'd be neat, but I can also just add them once we've merged this current one (with the minor changes I've pointed to above).

-Vincent

**Comment by NinedayWang** [comment]
@VHellendoorn Thanks for your review! I will deal with it as soon as possible :)

**Comment by urialon** [comment]
Hey @NinedayWang ,
It seems that Huggingface have released your PR! 
https://github.com/huggingface/transformers/releases/tag/v4.23.0

What are the next steps? Converting, pushing to the hub, and checking in the conversion scripts for future reference?

**Comment by NinedayWang** [comment]
@VHellendoorn @urialon 
Hi, I have made the suggested changes in 4018afe61d8cb9c3682fd29efe95c557cb437d61. I tested all PolyCoder models and got the expected results. Is it now possible to convert the models and push them to the hub?

File: Convert2HF/convert.sh
Comment (VHellendoorn) [inline Convert2HF/convert.sh:3]: Could you parameterize the model size in this command? Just adding `size=$1` as the first line and then replacing each occurrence of `0-4B` with `${size}` should work. If you want to make 0-4B the default size, you can use `size=${1:-0-4B}` instead.

File: Convert2HF/README.md
Comment (VHellendoorn) [inline Convert2HF/README.md]: How about this to emphasize how you can use the shell script. Note that this is contingent on making the model size a parameter, as suggested below.
```suggestion
You can use the `convert.sh` script to convert specified model to the HuggingFace format, using `./convert.sh 0-4B` (or pass a different model size). This script in turn invokes `convert_neox_pt_to_huggingface_neox.py`, which you can also call directly as follows:
```

File: Convert2HF/README.md
Comment (VHellendoorn) [inline Convert2HF/README.md:4]: Credit where credit is due :)
```suggestion
Shoutout to @NinedayWang for implementing this!

```

**Review by VHellendoorn — CHANGES_REQUESTED** [review]
Thanks @NinedayWang! I just left a few more small comments here and then it's good to merge.

File: Convert2HF/convert.sh
Comment (NinedayWang) [reply Convert2HF/convert.sh:3]: Done.

**Comment by urialon** [comment]
Hi @NinedayWang , 
Thank you for all the hard work!

Can you please go ahead and push all models to the Huggingface Hub?
@VHellendoorn will review the scripts, but in the meantime, we can push the models.

You can call them `PolyCoder-0.4B`, `PolyCoder-160M`, `PolyCoder-2.7B`.
I will add loading instructions to this README and details about this repo to the "model card" at the hub.

Thanks!

**Comment by NinedayWang** [comment]
> Hi @NinedayWang , Thank you for all the hard work!
> 
> Can you please go ahead and push all models to the Huggingface Hub? @VHellendoorn will review the scripts, but in the meantime, we can push the models.
> 
> You can call them `PolyCoder-0.4B`, `PolyCoder-160M`, `PolyCoder-2.7B`. I will add loading instructions to this README and details about this repo to the "model card" at the hub.
> 
> Thanks!

Yes of course I can push these models to the hub, but this seems like I would be the owner of them and your pull request would need my review. Maybe it would be better if you upload the models?

**Comment by urialon** [comment]
> > Hi @NinedayWang , Thank you for all the hard work!
> > Can you please go ahead and push all models to the Huggingface Hub? @VHellendoorn will review the scripts, but in the meantime, we can push the models.
> > You can call them `PolyCoder-0.4B`, `PolyCoder-160M`, `PolyCoder-2.7B`. I will add loading instructions to this README and details about this repo to the "model card" at the hub.
> > Thanks!
> 
> Yes of course I can push these models to the hub, but this seems like I would be the owner of them and your pull request would need my review. Maybe it would be better if you upload the models?

I think you can push the model to the hub and then enable "Enable access requests for this model" and give us access so we can edit the README of the model card. 
You can also push, and then transfer ownership to @VHellendoorn. @VHellendoorn what do you think?

**Comment by NinedayWang** [comment]
> > > Hi @NinedayWang , Thank you for all the hard work!
> > > Can you please go ahead and push all models to the Huggingface Hub? @VHellendoorn will review the scripts, but in the meantime, we can push the models.
> > > You can call them `PolyCoder-0.4B`, `PolyCoder-160M`, `PolyCoder-2.7B`. I will add loading instructions to this README and details about this repo to the "model card" at the hub.
> > > Thanks!
> > 
> > 
> > Yes of course I can push these models to the hub, but this seems like I would be the owner of them and your pull request would need my review. Maybe it would be better if you upload the models?
> 
> I think you can push the model to the hub and then enable "Enable access requests for this model" and give us access so we can edit the README of the model card. You can also push, and then transfer ownership to @VHellendoorn. @VHellendoorn what do you think?

It seems that neither method works：
1. If I enable "Enable access requests for this model", you need to perform an extra manual step to access the repository's content, and you still need to open a pull request and ask for my review.
2. Transferring ownership does not work for transferring a repository from user A to user B, referring to [this document](https://huggingface.co/docs/hub/repositories-settings#renaming-or-transferring-a-repo).

I've pushed these models to the hub ([PolyCoder-0.4B](https://huggingface.co/NinedayWang/PolyCoder-0.4B), [PolyCoder-160M](https://huggingface.co/NinedayWang/PolyCoder-160M), [PolyCoder-2.7B](https://huggingface.co/NinedayWang/PolyCoder-2.7B)), maybe you can download and re-upload them to your account, then I'll delete them in my space.

**Comment by urialon** [comment]
Thanks a lot @NinedayWang !
I created a pull request and added a model card.

I will also update the README of this repo.

**Pull request merged**
